lib/AST/ASTContext.cpp

```diff
@@ -4940,3 +4940,195 @@ bool ASTContext::isASCIIString(StringRef s) const {
+        existentialSig, funcTy->getResult(), skipAssocTypes);
+
+    auto kind = inputKind.flip();
+    kind |= resultKind;
+    return kind;
+  }
+
+  // Metatypes preserve variance.
+  if (auto metaTy = type->getAs<MetatypeType>()) {
+    return findExistentialSelfReferences(
+        existentialSig, metaTy->getInstanceType(), skipAssocTypes);
+  }
+
+  // Optionals preserve variance.
+  if (auto optType = type->getOptionalObjectType()) {
+    return findExistentialSelfReferences(existentialSig, optType,
+                                         skipAssocTypes);
+  }
+
+  // DynamicSelfType preserves variance.
+  // FIXME: This shouldn't ever appear in protocol requirement
+  // signatures.
+  if (auto selfType = type->getAs<DynamicSelfType>()) {
+    return findExistentialSelfReferences(
+        existentialSig, selfType->getSelfType(), skipAssocTypes);
+  }
+
+  // Bound generic types are invariant.
+  if (auto boundGenericType = type->getAs<BoundGenericType>()) {
+    for (auto paramType : boundGenericType->getGenericArgs()) {
+      if (findExistentialSelfReferences(existentialSig, paramType,
+                                        skipAssocTypes)) {
+        return SelfReferenceKind::Other();
+      }
+    }
+  }
+
+  const auto selfTy = existentialSig->getGenericParams().front();
+
+  // A direct reference to 'Self' is covariant.
+  if (selfTy->isEqual(type))
+    return SelfReferenceKind::Result();
+
+  if (skipAssocTypes)
+    return SelfReferenceKind::None();
+
+  // Special handling for associated types.
+  if (type->is<DependentMemberType>() &&
+      selfTy->isEqual(type->getRootGenericParam())) {
+    const auto typeInContext = existentialSig->getCanonicalTypeInContext(type);
+
+    // A direct reference to 'Self' is covariant.
+    if (selfTy->isEqual(typeInContext))
+      return SelfReferenceKind::Result();
+
+    if (typeInContext->hasTypeParameter())
+      return SelfReferenceKind::Other();
+  }
+
+  return SelfReferenceKind::None();
+}
+
+/// Find Self references in a generic signature's same-type requirements.
+static SelfReferenceKind
+findExistentialSelfReferences(GenericSignature existentialSig,
+                              GenericSignature targetSig) {
+  if (!targetSig)
+    return SelfReferenceKind::None();
+
+  const auto selfTy = existentialSig->getGenericParams().front();
+  for (const auto &req : targetSig->getRequirements()) {
+    if (req.getKind() != RequirementKind::SameType)
+      continue;
+
+    if (req.getFirstType()->isEqual(selfTy) ||
+        req.getSecondType()->isEqual(selfTy))
+      return SelfReferenceKind::Requirement();
+  }
+
+  return SelfReferenceKind::None();
+}
+
+SelfReferenceKind ASTContext::findExistentialSelfReferences(
+    Type existential, const ValueDecl *value, bool allowCovariantParameters,
+    bool skipAssocTypes) const {
+  assert(existential->isExistentialType());
+  const auto sig = existential->getASTContext().getOpenedArchetypeSignature(
```

Comment:
@slavapestov Is there a reason to build and store a whole new signature for a `ProtocolType` in  `getOpenedArchetypeSignature`? It seems to be identical to the protocol's own canonical generic signature.

Reply:
https://github.com/apple/swift/pull/33871